### PR TITLE
Polished Section 2 Creeps

### PR DESCRIPTION
**Pull request by Shushishtok**
- Sniper's bot will now keep himself behind enemy lines to make it more realistic. If all of his allies fall, he will fall back to an anchor point, and will regroup with his new allies, staying behind them.
- Sniper now has +100 damage (down from +600) to make not instantly explode any creep he denies. However, to make him more consistent with denies, he also now has +50 attack speed. Those values are removed when Sniper fights you in the last phase of the section.
- When the player hits an appropriate creep (enemies in last hit stage, allies in deny stage, none in breathe fire stage), and the creep stays alive with less than 7% of its health, a small "early!" message will show up above the creep, to hint that the player was attacking too early.
- When the player "misses" a last hit (<300 distance from enemy that died not thanks to the player), a small "miss" message will show up above the creep.
- Added two new util functions for calculation distance and direction. Nothing special, I just don't want to repeat myself anymore.
- Other slight fixes to behavior and general polish.

Fixes https://github.com/ModDota/dota-tutorial/issues/267

### CODE REVIEW

File: game/scripts/vscripts/util.ts [hunk 1]
```diff
@@ -507,3 +507,22 @@ export function setRespawnSettings(respawnLocation: Vector, respawnTime: number)
         }
     }, GameRules.Addon)
 }
+
+
+/**
+ * Calculates the distance in 2D between two points.
+ * @returns The distance between the two points.
+ */
+export function CalculateDistance2D(point1: Vector, point2: Vector): number {
```
Comment (RobinKa) [inline game/scripts/vscripts/util.ts:516]: Just `Distance2D` imo

File: game/scripts/vscripts/util.ts [hunk 2]
```diff
@@ -507,3 +507,22 @@ export function setRespawnSettings(respawnLocation: Vector, respawnTime: number)
...
+    return ((point1 - point2) as Vector).Length2D()
+}
+
+/**
+ * Calculates the direction from the first position to the second.
+ * @param origin_pos The position where the vector should originate from.
+ * @param towards_pos The position where the vector should point.
+ * @returns The normalized vector pointing towards the second vector argument.
+ */
+export function CalculateDirectionToPosition(origin_pos: Vector, towards_pos: Vector): Vector {
```
Comment (RobinKa) [inline game/scripts/vscripts/util.ts:526]: Also here `DirectionToPosition` or `DirectionBetween`?

File: game/scripts/vscripts/modifiers/modifier_sniper_deny_chapter2_creeps.ts [hunk 3]
```diff
@@ -49,25 +46,58 @@ export class modifier_sniper_deny_chapter2_creeps extends BaseModifier {
...
+                            {
+                                OrderType: UnitOrder.MOVE_TO_POSITION,
+                                UnitIndex: this.GetParent().entindex(),
+                                Position: position
+                            }
+                        )
+                    }
+                }
+                else {
+
```
Comment (RobinKa) [inline game/scripts/vscripts/modifiers/modifier_sniper_deny_chapter2_creeps.ts:75]:
```ts
else if (...) {
...
} else {
...
}
```

File: game/scripts/vscripts/modifiers/modifier_sniper_deny_chapter2_creeps.ts [hunk 4]
```diff
@@ -49,25 +46,58 @@ export class modifier_sniper_deny_chapter2_creeps extends BaseModifier {
...
+                if (alliedCreeps.length > 0) {
+                    const closestCreep = alliedCreeps[0]
+
+                    // Calculate the position Sniper should move behind enemy lines
+                    const direction = CalculateDirectionToPosition(closestCreep.GetAbsOrigin(), this.anchorPoint)
+                    const position = (closestCreep.GetAbsOrigin() + direction * 300) as Vector
+
+                    if (CalculateDistance2D(this.GetParent().GetAbsOrigin(), position) > 100) {
+                        ExecuteOrderFromTable(
+                            {
```
Comment (RobinKa) [inline game/scripts/vscripts/modifiers/modifier_sniper_deny_chapter2_creeps.ts:66]: `({` and `})`

File: game/scripts/vscripts/modifiers/modifier_dk_last_hit_chapter2_creeps.ts [hunk 5]
```diff
@@ -26,7 +26,35 @@ export class modifier_dk_last_hit_chapter2_creeps extends BaseModifier {
...
+
+    OnAttackLanded(event: ModifierAttackEvent): void {
+        if (!IsServer()) return;
+        if (event.attacker != this.GetParent()) return;
+        if (!isCustomLaneCreepUnit(event.target)) return
+
+        if (this.currentStage === LastHitStages.LAST_HIT && event.target.GetTeamNumber() === this.GetParent().GetTeamNumber()) {
+            return
+        }
+        else if (this.currentStage === LastHitStages.LAST_HIT_BREATHE_FIRE) {
```
Comment (RobinKa) [inline game/scripts/vscripts/modifiers/modifier_dk_last_hit_chapter2_creeps.ts:42]: `} else if (...) {`

File: game/scripts/vscripts/modifiers/modifier_dk_last_hit_chapter2_creeps.ts [hunk 6]
```diff
@@ -26,7 +26,35 @@ export class modifier_dk_last_hit_chapter2_creeps extends BaseModifier {
...
+            return
+        }
+        else if (this.currentStage === LastHitStages.LAST_HIT_DENY && event.target.GetTeamNumber() !== this.GetParent().GetTeamNumber()) {
```
Comment (RobinKa) [inline game/scripts/vscripts/modifiers/modifier_dk_last_hit_chapter2_creeps.ts:45]: `} else if (...) {`

File: game/scripts/vscripts/modifiers/modifier_sniper_deny_chapter2_creeps.ts [hunk 7]
```diff
@@ -49,25 +46,58 @@ export class modifier_sniper_deny_chapter2_creeps extends BaseModifier {
...
+
+                    if (CalculateDistance2D(this.GetParent().GetAbsOrigin(), this.anchorPoint) > 100) {
+                        ExecuteOrderFromTable({
+                            OrderType: UnitOrder.MOVE_TO_POSITION,
                             UnitIndex: this.GetParent().entindex(),
-                            TargetIndex: closestCreep.entindex(),
+                            Position: this.anchorPoint
                         })
+                    }
+                    else {
```
Comment (RobinKa) [inline game/scripts/vscripts/modifiers/modifier_sniper_deny_chapter2_creeps.ts:83]: `} else {`

**Review by RobinKa — COMMENTED** [review]
Looks good

File: game/scripts/vscripts/util.ts (hunk 2 above)
Comment (Shushishtok) [reply game/scripts/vscripts/util.ts:526]: DirectionToPosition is correct.

**Pull request merged**